rustls-symcrypt/src/verify.rs

```diff
@@ -0,0 +1,384 @@
+use pkcs1::RsaPublicKey as AsnRsaPublicKey;
+use rustls::crypto::WebPkiSupportedAlgorithms;
+use rustls::pki_types::{AlgorithmIdentifier, InvalidSignature, SignatureVerificationAlgorithm};
+use rustls::SignatureScheme;
+use symcrypt::ecc::{CurveType, EcKey, EcKeyUsage};
+use symcrypt::hash::{sha256, sha384, sha512, HashAlgorithm};
+use symcrypt::rsa::{RsaKey, RsaKeyUsage};
+use webpki::alg_id::{self};
+
+/// Rsa signatures from the wire will come in the following ASN1 format:
+/// RSAPublicKey ::= SEQUENCE {
+///     modulus           INTEGER,  -- n
+///     publicExponent    INTEGER   -- e
+/// }
+fn extract_rsa_public_key(pub_key: &[u8]) -> Result<(Vec<u8>, Vec<u8>), InvalidSignature> {
+    let key = AsnRsaPublicKey::try_from(pub_key).map_err(|_| InvalidSignature)?;
+
+    let modulus = key.modulus.as_bytes().to_vec();
+    let exponent = key.public_exponent.as_bytes().to_vec();
+
+    Ok((modulus, exponent))
+}
+
+// NistP256, NistP384, and NistP521 have a prepended Legacy byte that needs to be removed
+fn extract_ecc_public_key(pub_key: &[u8], curve_type: CurveType) -> Vec<u8> {
+    if curve_type == CurveType::Curve25519 {
+        // x25519 does not have the legacy byte prepended
+        return pub_key[..].to_vec();
+    }
+    // remove prepended 0x04 legacy byte
+
+
+    // remove only if 0x04 is there 
+    pub_key[1..].to_vec()
+}
+
+/// Ecc signatures from the wire will come in the following ASN1 format:
+/// ECDSASignature ::= SEQUENCE {
+///     r INTEGER,
+///     s INTEGER
+/// }
+/// SymCrypt expects a concatenated r+s with leading padding and leading 0's for both r and s to be removed
+fn extract_ecc_signature(signature: &[u8], curve: CurveType) -> Result<Vec<u8>, InvalidSignature> {
+    // We use pkcs1::RsaPublicKey because the underlying ASN1 format between an RSA public key and
+    // an ECC signgature is the same.
+    let signature = AsnRsaPublicKey::try_from(signature).map_err(|_| InvalidSignature)?;
+    
+    // r and s will both be the curve size / 2.
+    let component_length = curve.get_size() / 2;
```

Comment:
> / 2; [](http://example.com/codeflow?start=43&length=5)

// Per SymCrypt docs, the size of the signature will be 2 x the size of the private key.
        let signature_size = self.get_size_of_private_key() * 2;

So I would expect the component_length to be self.get_size_of_private_key()

Reply:
Fixed